contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -1166,3 +1186,239 @@ has_ilike_node_and_ci_as_coll(Node *expr)
 	}
 	return false;
 }
+
+/*
+ * For a given string and position in UTF-16 and return
+ * the corresponding position in UTF-8 string. UTF-16
+ * string considers surrogate pairs as two chars while
+ * in UTF-8 they are 1, which is why we need to translate
+ */
+
+static int32_t
+translate_char_pos(const char* str,		/* UTF-8 string */
+				   int32_t str_len,		/* length of UTF-8 string */
+				   const UChar* str_utf16, 		/* UTF-16 string */
+				   int32_t u16_len,		/* length of UTF-16 string */
+				   int32_t u16_pos,		/* position to translare in UTF-16 string */
+				   const char **p_str)		/* character at same position in UTF-8 string */
+{
+	UChar32 c;
+	int32_t u16_idx = 0;
+	int32_t out_pos = 0;
+	int32_t u8_offset = 0;
+
+	Assert (GetDatabaseEncoding() == PG_UTF8);
+
+	/* for UTF-8, use ICU macros instead of calling pg_mblen() */
+	while (u16_idx < u16_pos)
+	{
+#ifdef USE_ICU
+		U16_NEXT(str_utf16, u16_idx, u16_len, c);
+		U8_NEXT(str, u8_offset, str_len, c);
+		out_pos++;
+#else
+	ereport(ERROR,
+			(errcode(ERRCODE_EXTERNAL_ROUTINE_EXCEPTION),
+				errmsg("This function requires ICU library, which is not available")));
```

Comment:
"This function" is somewhat ambiguous, let's be more specific in error message here and everywhere

Reply:
done